R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
-    # https://msalganik.wordpress.com/2017/01/21/making-sense-of-the-rlnorm-function-in-r/comment-page-1/
-    mean_log <- log(estimate^2 / sqrt(estimate^2 + variance))
-    sd_log <- sqrt(log(1 + (variance / estimate^2)))
-    ci <- purrr::map2(mean_log, sd_log, ~ stats::qlnorm(c(q_low, q_high), .x, .y))
-  } else if (dist == "poisson") {
-    ci <- purrr::map2(estimate, denominator, ~ get_ci_pois(q_low, q_high, .x, .y))
-  }
-
-  # wrangle output
-  result <- ci %>%
-    data.frame() %>%
-    data.table::transpose() %>%
-    dplyr::rename(lower = 1, upper = 2)
+  q_upper <- 1 - (alpha / 2)
+  q_lower <- alpha / 2
+
+  result <- data.frame(x = x, y = y) %>%
+    dplyr::mutate(
+      lower = stats::qchisq(q_lower, df = 2 * x) / (2 * y),
+      upper = stats::qchisq(q_upper, df = 2 * (x + 1)) / (2 * y)
+    ) %>%
+    dplyr::select(.data$lower, .data$upper)
 
   return(result)
 }
 
 
-#' Replicate `stats::qpois` using `stats::qchisq`
+#' Calculate a gamma distribution confidence interval
+#'
+#' @description Derive an approximate gamma distribution confidence interval
+#' using [stats::qchisq()].
+#'
+#' @param interval A scalar between 0 and 1, indicating the width of the
+#' interval. For example, use `0.95` to calculate a 95% confidence interval.
+#' @param estimate A numeric vector containing the point estimates.
+#' @param weights A list of numeric vectors containing the weights.
+#' @param variance A numeric vector containing the variances.
+#' @param method A string to indicate which method to use to calculate the
+#' confidence interval. The default is `"tcz06"` which refers to the method
+#' proposed by Tiwari, Clegg, and Zou (2006). Use `"ff97"` for a more
+#' conservative confidence interval proposed by Fay and Feuer (1997).
 #'
-#' @description This function implements the Poisson distribution using the
-#' chi-squared distribution to allow for the continuous extension of integer
-#' variable (`estimate`). Read p.9385 of the [STDRATE Procedure](https://support.sas.com/documentation/onlinedoc/stat/151/stdrate.pdf)
-#' for more information.
+#' @details This function assumes that the values of the input arguments
+#' `estimate`, `weights`, and `variance` are aligned and that they are the same
+#' length.
 #'
-#' @param estimate A scalar, must be a whole number (e.g. number of events)
-#' @param denominator A scalar (e.g. population-time)
-#' @param q_low A scalar, indicating the lower limit
-#' @param q_high A scalar, indicating the upper limit
+#' This function is used to add confidence intervals in [get_ds_rt()].
 #'
-#' @return A numeric vector containing the lower and upper limits
+#' @return A data frame with columns `lower` and `upper` which contain the lower
+#' and upper limits of a confidence interval respectively.
+#'
+#' @references \href{http://documentation.sas.com/doc/en/pgmsascdc/9.4_3.4/statug/statug_stdrate_details.htm}{The STDRATE Procedure}
```

Comment:
I think the link should maybe go to the page with the gamma distribution specifically?

i.e. http://documentation.sas.com/doc/en/pgmsascdc/9.4_3.4/statug/statug_stdrate_details14.htm#statug_stdrate002267

Reply:
Or maybe keep this link here as-is (since the other distributions, which are used in other metrics just link to the top level page in the SAS manual). and then add this new link (directly to the gamma dist for DSR) as a link when the `method` parameter is explained.